exchange/bitswap/bitswap.go

```diff
@@ -375,6 +373,10 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 			k := b.Cid()
 			log.Event(ctx, "Bitswap.GetBlockRequest.End", k)
 
+			for _, ses := range bs.SessionsForBlock(k) {
+				ses.receiveBlockFrom(p, b)
```

Comment:
> That race condition exists in the code today

The first one already exists but the second one is a bit different and has a much larger window. To reproduce:

1. Try fetching 3000 nodes that can't be found on the network.
2. Manually add 100th node.
3. Manually add the rest.

The operation will never complete because, when added, the 100th node was neither in the interest set nor in the `liveWants` set. The quick fix is to add an `allWants` set and use that instead of of checking `liveWants` and `interest`.

---

However, *all* these race conditions like this could be fixed by *some* centralized notification system (doesn't have to be built into the blockstore).